File: webflux-dialogue/src/main/java/com/study/webflux/rag/infrastructure/config/MonitoringConfiguration.java
package com.study.webflux.rag.infrastructure.config;

import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Primary;

import com.study.webflux.rag.application.monitoring.LoggingPipelineMetricsReporter;
import com.study.webflux.rag.application.monitoring.PersistentPipelineMetricsReporter;
import com.study.webflux.rag.application.monitoring.PipelineMetricsReporter;
import com.study.webflux.rag.domain.port.out.PerformanceMetricsRepository;
import com.study.webflux.rag.domain.port.out.UsageAnalyticsRepository;

@Configuration
public class MonitoringConfiguration {

	@Bean
	@Primary
	@ConditionalOnProperty(name = "monitoring.persistent.enabled", havingValue = "true", matchIfMissing = true)
	public PipelineMetricsReporter persistentPipelineMetricsReporter(
		PerformanceMetricsRepository performanceMetricsRepository,
		UsageAnalyticsRepository usageAnalyticsRepository,
		LoggingPipelineMetricsReporter loggingReporter) {
		return new PersistentPipelineMetricsReporter(
			performanceMetricsRepository,
			usageAnalyticsRepository,
			loggingReporter);
	}
}

File: backend/monitoring/src/main/kotlin/com/miyou/app/monitoring/config/MonitoringConfiguration.kt
package com.miyou.app.monitoring.config

import com.miyou.app.monitoring.adapter.LoggingPipelineMetricsReporter
import com.miyou.app.monitoring.adapter.StructuredLogPipelineMetricsReporter
import com.miyou.app.monitoring.micrometer.CompositePipelineMetricsReporter
import com.miyou.app.monitoring.micrometer.MicrometerPipelineMetricsReporter
import com.miyou.app.monitoring.port.PipelineMetricsReporter
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Primary

/**
 * 모니터링 파이프라인 보고자 구성을 정의합니다.
 *
 * 건당 상세(프롬프트/RAG/LLM 응답)는 [StructuredLogPipelineMetricsReporter]가 전용
 * 로그 파일로, 기간별 집계(비용/트래픽/병목)는 [MicrometerPipelineMetricsReporter]가
 * Prometheus로 각각 전담한다. Mongo 적재는 하지 않는다.
 */
@Configuration
class MonitoringConfiguration {
    @Bean
    @Primary
    fun pipelineMetricsReporter(
        loggingReporter: LoggingPipelineMetricsReporter,
        structuredLogReporter: StructuredLogPipelineMetricsReporter,
        micrometerReporter: MicrometerPipelineMetricsReporter,
    ): PipelineMetricsReporter =
        CompositePipelineMetricsReporter(
            listOf(
                loggingReporter,
                structuredLogReporter,
                micrometerReporter,
            ),
        )
}
